In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
import numpy as np

In [2]:
treaty_message_table = pd.read_csv('treaty_message_table')

In [3]:
treaty_message_table

Unnamed: 0                                         treaty  \
0             0            the nuclear nonproliferation treaty   
1             1                    the nonproliferation treaty   
2             2                      the north atlantic treaty   
3             3                           the new start treaty   
4             4    the intermediaterange nuclear forces treaty   
..          ...                                            ...   
127         127                          the arms trade treaty   
128         128                        the panama canal treaty   
129         129                        the iran nuclear treaty   
130         130                                the oslo treaty   
131         131  the organization's founding washington treaty   

     most common message  
0               OLD Iran  
1               OLD Iran  
2    OLD Missile Defense  
3     OLD Arms Reduction  
4     OLD Arms Reduction  
..                   ...  
127  OLD Missile Defense  
128             OLD Iran  
129    OLD Proliferation  
130    OLD Proliferation  
131             OLD Iran  

[132 rows x 3 columns]

In [4]:
merged_df = pd.read_csv('merged_df')

In [5]:
merged_df.head()

Unnamed: 0     artdate  Article ID                             Headline  \
0           0  2018-01-10     3759306  North Korea makes deals and threats   
1           1  2018-01-10     3759306  North Korea makes deals and threats   
2           2  2018-01-10     3759306  North Korea makes deals and threats   
3           3  2018-01-10     3759306  North Korea makes deals and threats   
4           4  2018-01-10     3759306  North Korea makes deals and threats   

   Quote ID Article Status Article Position           Messages  \
0   7023849           News          Neutral  OLD Proliferation   
1   7023842           News          Neutral  OLD Proliferation   
2   7023839           News          Neutral  OLD Proliferation   
3   7023833           News          Neutral  OLD Proliferation   
4   7023846           News          Neutral  OLD Proliferation   

                        Submessages Quote Position  ... Article Issues  \
0  OLD State Level Nuclear Programs        Neutral  ...            NaN   
1  OLD State Level Nuclear Programs       Negative  ...            NaN   
2  OLD State Level Nuclear Programs       Negative  ...            NaN   
3  OLD State Level Nuclear Programs       Negative  ...            NaN   
4  OLD State Level Nuclear Programs       Positive  ...            NaN   

  Custom Group Media Group                                            Content  \
0          NaN         NaN  North Korea makes deals and threats\r\n\r\nMed...   
1          NaN         NaN  North Korea makes deals and threats\r\n\r\nMed...   
2          NaN         NaN  North Korea makes deals and threats\r\n\r\nMed...   
3          NaN         NaN  North Korea makes deals and threats\r\n\r\nMed...   
4          NaN         NaN  North Korea makes deals and threats\r\n\r\nMed...   

          Author          Published Date  \
0  Fifield, Anna  2018-01-10 00:00:00+00   
1  Fifield, Anna  2018-01-10 00:00:00+00   
2  Fifield, Anna  2018-01-10 00:00:00+00   
3  Fifield, Anna  2018-01-10 00:00:00+00   
4  Fifield, Anna  2018-01-10 00:00:00+00   

                                   First_10%_Content  \
0  SEOUL - North Korea's representatives assured ...   
1  SEOUL - North Korea's representatives assured ...   
2  SEOUL - North Korea's representatives assured ...   
3  SEOUL - North Korea's representatives assured ...   
4  SEOUL - North Korea's representatives assured ...   

                             First_10%_Content_Split Treaty_Instances  \
0  ['SEOUL', '-', 'North', "Korea's", 'representa...               []   
1  ['SEOUL', '-', 'North', "Korea's", 'representa...               []   
2  ['SEOUL', '-', 'North', "Korea's", 'representa...               []   
3  ['SEOUL', '-', 'North', "Korea's", 'representa...               []   
4  ['SEOUL', '-', 'North', "Korea's", 'representa...               []   

  Treaties2  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN  

[5 rows x 36 columns]

In [6]:
# extract articles with mentions of treaties
treaty_articles = merged_df.dropna(subset=['Treaties2'])

In [7]:
treaty_articles = treaty_articles.loc[treaty_articles['Treaties2'].isin(treaty_message_table['treaty'])]

In [8]:
predicted_message = []
for i in np.arange(len(treaty_articles)):
    # find treaty in treaty_message_table, append most common message
    treaty = treaty_articles.reset_index()['Treaties2'][i]
    treaty_message = treaty_message_table.loc[treaty_message_table['treaty'] == treaty]['most common message'].item()
    predicted_message.append(treaty_message)
    

In [9]:
true_predicted_message_table = treaty_articles[['Messages']].reset_index(drop=True)
true_predicted_message_table['predicted_message'] = predicted_message
true_predicted_message_table = true_predicted_message_table.rename(columns= {'Messages': 'true_message'})

In [10]:
true_predicted_message_table

true_message    predicted_message
0       OLD Arms Reduction   OLD Arms Reduction
1       OLD Arms Reduction   OLD Arms Reduction
2       OLD Arms Reduction   OLD Arms Reduction
3        OLD Proliferation             OLD Iran
4        OLD Proliferation             OLD Iran
...                    ...                  ...
10584   OLD Arms Reduction   OLD Arms Reduction
10585    OLD Proliferation   OLD Arms Reduction
10586  OLD Missile Defense  OLD Missile Defense
10587   OLD Arms Reduction  OLD Missile Defense
10588  OLD Missile Defense  OLD Missile Defense

[10589 rows x 2 columns]

In [11]:
correct = sum(true_predicted_message_table['true_message'] == true_predicted_message_table['predicted_message'])
total = len(true_predicted_message_table)
accuracy = correct/total
accuracy

0.5761639437151761

# relationship between Treaty and Topic (submessages)

In [12]:
len(merged_df['Treaties2'].value_counts())

134

In [13]:
submessages_treates = merged_df[['Treaties2', 'Submessages']]

In [14]:
merged_df['Submessages'].value_counts()

OLD Negotiations                        87980
OLD State Level Nuclear Programs        73337
OLD Sanctions                           29007
OLD Iranian Domestic Politics           23389
OLD Deployment and Testing              17811
OLD Russia Reset                        10396
OLD Domestic Programs                    7774
OLD European/Russian Missile Defense     7164
OLD  Politics                            5967
OLD US Arsenal                           4994
OLD Covert Ops                           4368
OLD Iran Rapid Response                  4208
OLD Iran                                 4190
OLD Loose Nuclear Materials              3064
OLD Arms Reduction                       2754
OLD START Treaty                         2681
OLD Modernization                        2160
OLD Nukes Budget Campaign                1620
OLD Terrorist Threat                     1565
OLD Foreign Policy Narrative             1533
OLD Other                                 779
OLD NPT                           

In [15]:
def submessage_treaty_table(treaty, topic_column):
    table =  merged_df.loc[merged_df['Treaties2'] == treaty].groupby(by = topic_column).count()
    return table.reset_index().iloc[:,0:2].rename(columns = {'Unnamed: 0' : 'count'})

In [16]:
submessage_treaty_table('the nuclear nonproliferation treaty', 'Submessages')

Submessages  count
0                                    NPT      1
1                          OLD  Politics     11
2                    OLD AQ Khan Network      2
3                     OLD Arms Reduction    117
4                               OLD CTBT      5
5                         OLD Covert Ops     16
6             OLD Deployment and Testing     48
7                  OLD Domestic Programs      6
8   OLD European/Russian Missile Defense     22
9    OLD Fissile Material Cut Off Treaty      3
10          OLD Foreign Policy Narrative      4
11                              OLD Iran     25
12               OLD Iran Rapid Response     25
13         OLD Iranian Domestic Politics    630
14           OLD Loose Nuclear Materials     16
15                     OLD Modernization      3
16                               OLD NPT    272
17                      OLD Negotiations   1520
18                 OLD Non-Proliferation      4
19             OLD Nukes Budget Campaign      4
20                             OLD Other      3
21                     OLD Proliferation      6
22                      OLD Russia Reset     13
23                      OLD START Treaty      6
24                         OLD Sanctions    221
25      OLD State Level Nuclear Programs    991
26             OLD Tactical Nukes Treaty     14
27                  OLD Terrorist Threat     15
28                        OLD US Arsenal     39
29                    US Nuclear Arsenal      1

In [17]:
treaties_unique = pd.DataFrame(merged_df['Treaties2'].value_counts())
treaties_unique#.loc[treaties_unique['Treaties2'] > 50]

Treaties2
the nuclear nonproliferation treaty                 4043
the nonproliferation treaty                          879
the north atlantic treaty                            741
the new start treaty                                 545
the intermediaterange nuclear forces treaty          543
...                                                  ...
the panama canal treaty                                1
the iran nuclear treaty                                1
the oslo treaty                                        1
the organization's founding washington treaty          1
the alliance’s founding washington treaty              1

[134 rows x 1 columns]

# find proportion of articles with the given treaty in the different Submessages


In [18]:
def submessage_treaty_props(treaty, topic_column):
    table =  merged_df.loc[merged_df['Treaties2'] == treaty].groupby(by = topic_column).count()
    table = table.reset_index().iloc[:,0:2].rename(columns = {'Unnamed: 0' : 'count'})
    total = sum(table['count'])
    props = []
    for i in np.arange(len(table)):
        props.append(table['count'][i] / total)
    table['props'] = props
    table['treaty'] = treaty
    return table[['treaty', topic_column, 'props']]



In [19]:
# treaties most --> least common
treaties_list = treaties_unique.index

In [20]:
treaty_submessage_table = submessage_treaty_props(treaties_list[0], 'Submessages')
treaty_submessage_table= pd.pivot_table(treaty_submessage_table, values = 'props', index = ['treaty'], columns = ['Submessages'])

for i in treaties_list[1:]:
    table = submessage_treaty_props(i, 'Submessages')
    table = pd.pivot_table(table, values = 'props', index = ['treaty'], columns = ['Submessages'])
    treaty_submessage_table = pd.concat([treaty_submessage_table, table])

In [21]:
pd.set_option('display.max_rows', len(treaty_submessage_table))
treaty_submessage_table

Submessages                                              NPT  OLD  Politics  \
treaty                                                                        
the nuclear nonproliferation treaty                 0.000247       0.002721   
the nonproliferation treaty                              NaN       0.003417   
the north atlantic treaty                                NaN       0.096467   
the new start treaty                                     NaN       0.031193   
the intermediaterange nuclear forces treaty              NaN       0.037244   
the treaty                                               NaN       0.017045   
the comprehensive test ban treaty                        NaN       0.002959   
the antiballistic missile treaty                         NaN       0.043137   
the inf treaty                                           NaN       0.004878   
the comprehensive nucleartestban treaty                  NaN            NaN   
the 1968 nuclear nonproliferation treaty                 NaN       0.035928   
the 1970 nuclear nonproliferation treaty                 NaN            NaN   
the strategic arms reduction treaty                      NaN       0.055556   
the nuclear non proliferation treaty                     NaN            NaN   
the comprehensive testban treaty                         NaN            NaN   
the 1972 antiballistic missile treaty                    NaN            NaN   
the international nuclear nonproliferation treaty        NaN            NaN   
the start treaty                                         NaN            NaN   
the intermediate nuclear forces treaty                   NaN            NaN   
the 2015 nuclear nonproliferation treaty                 NaN            NaN   
the nato treaty                                          NaN            NaN   
the 2010 new start treaty                                NaN            NaN   
the limited test ban treaty                              NaN            NaN   
the abm treaty                                           NaN            NaN   
the non proliferation treaty                             NaN            NaN   
the north american treaty                                NaN            NaN   
the npt (nonproliferation treaty                         NaN            NaN   
the global nonproliferation treaty                       NaN            NaN   
the nuclearnonproliferation treaty                       NaN            NaN   
the 1970 nonproliferation treaty                         NaN            NaN   
the [nuclear nonproliferation treaty                     NaN            NaN   
the nuclear test ban treaty                              NaN            NaN   
the washington treaty                                    NaN       0.100000   
the comprehensive nuclear testban treaty                 NaN            NaN   
the strategic offensive reductions treaty                NaN            NaN   
the u.n.'s nuclear nonproliferation treaty               NaN            NaN   
the sea treaty                                           NaN            NaN   
the fissile material cutoff treaty                       NaN            NaN   
the nowdefunct antiballistic missile treaty              NaN            NaN   
the 1968 nonproliferation treaty                         NaN            NaN   
the parliament's move.=========================...       NaN            NaN   
the comprehensive nucleartest ban treaty                 NaN            NaN   
the international nonproliferation treaty                NaN            NaN   
the global nuclear nonproliferation treaty               NaN            NaN   
the 1987 inf treaty                                      NaN       0.047619   
the 189nation nuclear nonproliferation treaty            NaN            NaN   
the 1967 outer space treaty                              NaN            NaN   
the u.s.russia new start treaty                          NaN            NaN   
the outer space treaty                            

In [22]:
# table with treaty & most common message (highest prop)
most_common_submessage = []
for i in np.arange(len(treaty_submessage_table)):
    most_common_submessage.append(pd.DataFrame(treaty_submessage_table.iloc[i, :]).idxmax()[0])

In [23]:
treaty_submessage_table['most common submessage'] = most_common_submessage

In [24]:
treaty_submessage_table = treaty_submessage_table.reset_index()[['treaty', 'most common submessage']]
treaty_submessage_table = treaty_submessage_table.rename_axis(None, axis=1)

In [25]:
treaty_submessage_table

treaty  \
0                  the nuclear nonproliferation treaty   
1                          the nonproliferation treaty   
2                            the north atlantic treaty   
3                                 the new start treaty   
4          the intermediaterange nuclear forces treaty   
5                                           the treaty   
6                    the comprehensive test ban treaty   
7                     the antiballistic missile treaty   
8                                       the inf treaty   
9              the comprehensive nucleartestban treaty   
10            the 1968 nuclear nonproliferation treaty   
11            the 1970 nuclear nonproliferation treaty   
12                 the strategic arms reduction treaty   
13                the nuclear non proliferation treaty   
14                    the comprehensive testban treaty   
15               the 1972 antiballistic missile treaty   
16   the international nuclear nonproliferation treaty   
17                                    the start treaty   
18              the intermediate nuclear forces treaty   
19            the 2015 nuclear nonproliferation treaty   
20                                     the nato treaty   
21                           the 2010 new start treaty   
22                         the limited test ban treaty   
23                                      the abm treaty   
24                        the non proliferation treaty   
25                           the north american treaty   
26                    the npt (nonproliferation treaty   
27                  the global nonproliferation treaty   
28                  the nuclearnonproliferation treaty   
29                    the 1970 nonproliferation treaty   
30                the [nuclear nonproliferation treaty   
31                         the nuclear test ban treaty   
32                               the washington treaty   
33            the comprehensive nuclear testban treaty   
34           the strategic offensive reductions treaty   
35          the u.n.'s nuclear nonproliferation treaty   
36                                      the sea treaty   
37                  the fissile material cutoff treaty   
38         the nowdefunct antiballistic missile treaty   
39                    the 1968 nonproliferation treaty   
40   the parliament's move.========================...   
41            the comprehensive nucleartest ban treaty   
42           the international nonproliferation treaty   
43          the global nuclear nonproliferation treaty   
44                                 the 1987 inf treaty   
45       the 189nation nuclear nonproliferation treaty   
46                         the 1967 outer space treaty   
47                     the u.s.russia new start treaty   
48                              the outer space treaty   
49                       the threshold test ban treaty   
50            the npt (nuclear nonproliferation treaty   
51                the strategic arms limitation treaty   
52                 the intermediaterange forces treaty   
53                                the firstever treaty   
54      the decadesold nuclear nonproliferation treaty   
55                           the 2011 new start treaty   
56                          the 1992 open skies treaty   
57       the 47yearold nuclear nonproliferation treaty   
58                               the open skies treaty   
59        the landmark nuclear nonproliferation treaty   
60                  the rushedthrough new start treaty   
61                    the nuclear proliferation treaty   
62        the valuable nuclear nonproliferation treaty   
63                   the congressional test ban treaty   
64                                     the u.n. treaty   
65                                     the 1990 treaty   
66                                   the i.n.f. treaty   
67      the international antiballistic missile treaty   
68                      the next north atlanti

# accuracy of topic (submessages) classification based on treaty

In [26]:
# extract articles with mentions of treaties
treaty_articles = merged_df.dropna(subset=['Treaties2'])

In [27]:
treaty_articles = treaty_articles.loc[treaty_articles['Treaties2'].isin(treaty_submessage_table['treaty'])]

In [28]:
predicted_submessage = []
for i in np.arange(len(treaty_articles)):
    # find treaty in treaty_message_table, append most common message
    treaty = treaty_articles.reset_index()['Treaties2'][i]
    treaty_submessage = treaty_submessage_table.loc[treaty_submessage_table['treaty'] == treaty]['most common submessage'].item()
    predicted_submessage.append(treaty_submessage)
    

In [29]:
true_predicted_submessage_table = treaty_articles[['Submessages']].reset_index(drop=True)
true_predicted_submessage_table['predicted_submessage'] = predicted_submessage
true_predicted_submessage_table = true_predicted_submessage_table.rename(columns= {'Submessages': 'true_submessage'})

In [30]:
true_predicted_submessage_table

true_submessage                  predicted_submessage
0                    OLD Arms Reduction                    OLD Arms Reduction
1                    OLD Arms Reduction                    OLD Arms Reduction
2                    OLD Arms Reduction                    OLD Arms Reduction
3      OLD State Level Nuclear Programs                      OLD Negotiations
4      OLD State Level Nuclear Programs                      OLD Negotiations
...                                 ...                                   ...
10584                  OLD START Treaty                        OLD INF Treaty
10585  OLD State Level Nuclear Programs                        OLD INF Treaty
10586        OLD Deployment and Testing  OLD European/Russian Missile Defense
10587                    OLD INF Treaty  OLD European/Russian Missile Defense
10588        OLD Deployment and Testing  OLD European/Russian Missile Defense

[10589 rows x 2 columns]

In [31]:
correct = sum(true_predicted_submessage_table['true_submessage'] == true_predicted_submessage_table['predicted_submessage'])
total = len(true_predicted_submessage_table)
accuracy = correct/total
accuracy

0.4033430918878081

# only post april 2020 data

In [33]:
post_april_2020 = merged_df.loc[merged_df['Published Date'] >= '2020-04-00 00:00:00+00']

In [34]:
post_april_2020['Submessages'].unique()

array([nan, 'Loose Fissile Materials', 'US/NATO Missile Defense',
       'Conflict with North Korea', 'Health Issues',
       'OLD State Level Nuclear Programs', 'JCPOA', 'Iran Negotiations',
       'Spending and Modernization', 'US Nuclear Arsenal', 'NPT',
       'Worldwide Nukes Policy'], dtype=object)

find proportion of articles with the given treaty in the different Submessages


In [35]:
def submessage_treaty_props2(treaty, topic_column):
    table =  post_april_2020.loc[post_april_2020['Treaties2'] == treaty].groupby(by = topic_column).count()
    table = table.reset_index().iloc[:,0:2].rename(columns = {'Unnamed: 0' : 'count'})
    total = sum(table['count'])
    props = []
    for i in np.arange(len(table)):
        props.append(table['count'][i] / total)
    table['props'] = props
    table['treaty'] = treaty
    return table[['treaty', topic_column, 'props']]



In [36]:
treaty_submessage_table2 = submessage_treaty_props2(treaties_list[0], 'Submessages')
treaty_submessage_table2= pd.pivot_table(treaty_submessage_table2, values = 'props', index = ['treaty'], columns = ['Submessages'])

for i in treaties_list[1:]:
    table = submessage_treaty_props2(i, 'Submessages')
    table = pd.pivot_table(table, values = 'props', index = ['treaty'], columns = ['Submessages'])
    treaty_submessage_table2 = pd.concat([treaty_submessage_table2, table])

In [37]:
pd.set_option('display.max_rows', len(treaty_submessage_table2))
treaty_submessage_table2

Submessages                          NPT  US Nuclear Arsenal
treaty                                                      
the nuclear nonproliferation treaty  0.5                 0.5

In [38]:
# table with treaty & most common message (highest prop)
most_common_submessage2 = []
for i in np.arange(len(treaty_submessage_table2)):
    most_common_submessage2.append(pd.DataFrame(treaty_submessage_table2.iloc[i, :]).idxmax()[0])

In [39]:
treaty_submessage_table2['most common submessage'] = most_common_submessage2

In [40]:
treaty_submessage_table2 = treaty_submessage_table2.reset_index()[['treaty', 'most common submessage']]
treaty_submessage_table2 = treaty_submessage_table2.rename_axis(None, axis=1)

In [41]:
treaty_submessage_table2

treaty most common submessage
0  the nuclear nonproliferation treaty                    NPT

accuracy of topic (submessages) classification based on treaty

In [42]:
# extract articles with mentions of treaties
treaty_articles2 = post_april_2020.dropna(subset=['Treaties2'])

In [43]:
treaty_articles2 = treaty_articles2.loc[treaty_articles2['Treaties2'].isin(treaty_submessage_table2['treaty'])]

In [44]:
predicted_submessage2 = []
for i in np.arange(len(treaty_articles2)):
    # find treaty in treaty_message_table, append most common message
    treaty = treaty_articles2.reset_index()['Treaties2'][i]
    treaty_submessage = treaty_submessage_table2.loc[treaty_submessage_table2['treaty'] == treaty]['most common submessage'].item()
    predicted_submessage2.append(treaty_submessage)
    

In [45]:
true_predicted_submessage_table2 = treaty_articles2[['Submessages']].reset_index(drop=True)
true_predicted_submessage_table2['predicted_submessage'] = predicted_submessage2
true_predicted_submessage_table2 = true_predicted_submessage_table2.rename(columns= {'Submessages': 'true_submessage'})

In [46]:
true_predicted_submessage_table2

true_submessage predicted_submessage
0   US Nuclear Arsenal                  NPT
..                 ...                  ...

[2 rows x 2 columns]

In [47]:
correct = sum(true_predicted_submessage_table2['true_submessage'] == true_predicted_submessage_table2['predicted_submessage'])
total = len(true_predicted_submessage_table2)
accuracy = correct/total
accuracy

0.5

main findings
- treaties better for classifying message topics (than submessage topic categories)
    - accuracies:
        - presence of certain treaty (treaty, Messages) ~ 0.58 (entire dataset)
        - presence of certain treaty (treaty, Submessages) ~ 0.40 (entire dataset)
        - presence of certain treaty (treaty, Submessages) ~ 0.5 (articles after April 2020)